feat: optimize images on upload + bulk optimize/analyze CLI commands

aseemann opened the pull request:
Jira: [NEXT-99](https://jira.netresearch.de/browse/NEXT-99)

## Summary

Adds automatic image optimization for uploaded/replaced files and two bulk CLI commands for operating on existing images in the FAL index.

## Changes

### New features

- **`OptimizeOnUploadListener`** — PSR-14 event listener reacting to `AfterFileAddedEvent` and `AfterFileReplacedEvent`. Automatically runs `optipng`, `gifsicle`, or `jpegoptim` on supported uploads. Re-entrancy-safe (won't loop on its own `replaceFile` calls).
- **`nr:image:optimize` CLI** — walks the FAL index and lossy/losslessly optimizes PNG, GIF, and JPEG files across all TYPO3 storages. Supports `--dry-run`, `--storages`, `--jpeg-quality`, `--strip-metadata`.
- **`nr:image:analyze` CLI** — reports optimization potential using a heuristic (size + resolution) without invoking external tools. Fast even on large installations.
- **`ImageOptimizer` service** — shared backend used by both the listener and the CLI commands. Discovers tool binaries via `PATH` with `OPTIPNG_BIN` / `GIFSICLE_BIN` / `JPEGOPTIM_BIN` env overrides.

### CLI UX

- Upfront record count, `SymfonyStyle` progress bar with MB/GB savings totals
- Progress labels truncated to fit terminal width
- Streaming iteration (Generator) over `sys_file` to avoid loading the full index into memory
- Permission checks temporarily disabled per file then restored (works in CLI context without a BE user)

### Tooling

- Full unit test coverage for the new classes (100 new tests, all in `Tests/Unit/`)
- Line coverage on patch files: `OptimizeOnUploadListener` 100%, `AnalyzeImagesCommand` 100%, `AbstractImageCommand` 96%, `OptimizeImagesCommand` 96%, `ImageOptimizer` 96%

## Testing

```bash
# Unit tests
composer ci:test:php:unit

# Static analysis
composer ci:test:php:phpstan
composer ci:test:php:rector
composer ci:test:php:cgl

# Try the analyze command (no modifications)
./vendor/bin/typo3 nr:image:analyze --storages=1 --min-size=512000

# Try the optimize command (dry-run first!)
./vendor/bin/typo3 nr:image:optimize --dry-run --storages=1
./vendor/bin/typo3 nr:image:optimize --storages=1 --jpeg-quality=85 --strip-metadata
```

## Requirements

The auto-optimization on upload works silently if no optimizer binaries are installed. To actually optimize, install at least one of:

- `optipng` (Debian/Ubuntu: `apt install optipng`)
- `gifsicle` (`apt install gifsicle`)
- `jpegoptim` (`apt install jpegoptim`)

CybotTM commented on Classes/Command/OptimizeImagesCommand.php:
```diff
@@ -0,0 +1,269 @@
+
+        return $args;
+    }
+
```
Fixed in 333532c — OptimizeImagesCommand now delegates to ImageOptimizer service, eliminating all duplicated methods (resolveToolForExtension, resolveBinary, buildJpegoptimArgs).

CybotTM commented on Classes/Command/OptimizeImagesCommand.php:
```diff
@@ -0,0 +1,269 @@
+            $args[] = '--max=' . $q; // lossy, gewünschte Zielqualität
+        }
+
+        // jpegoptim schreibt in-place, -q reduziert Ausgabe
```
Fixed in 3d3e56f — all German comments and UI strings have been translated to English.

CybotTM commented on Classes/EventListener/OptimizeOnUploadListener.php:
```diff
@@ -0,0 +1,68 @@
+        $this->guard[$id] = true;
+        try {
+            $this->optimizer->optimize($file, false, null, false);
+        } finally {
```
Fixed — all three locations now save previousPermissions and restore in a finally block. See commits 1125afc, 333532c, and bc2a59a.

CybotTM commented on Classes/Command/OptimizeImagesCommand.php:
```diff
@@ -0,0 +1,269 @@
+                if ($dryRun) {
+                    $io->writeln(sprintf('[DRY] Würde %s mit %s optimieren', $file->getIdentifier(), $tool['name']));
+                    continue;
+                }
```
Fixed — loop bodies are now wrapped in try/finally blocks that always call $progress->advance(). See commits 333532c and bc2a59a.

CybotTM commented on Classes/Processor.php:
```diff
@@ -312,4 +295,14 @@ public function getLocker(string $key): LockingStrategyInterface
+        $serverRequest = ServerRequestFactory::fromGlobals();
+
+        $accept   = $serverRequest->getHeader('accept')[0] ?? '';
+        $variants = explode(',', $accept);
```
The Processor has been reverted to the main branch version which uses query parameter-based variant selection (skipWebP/skipAvif) instead of Accept header parsing. The determineBrowserSupport method was removed entirely in ec0c1fb.

CybotTM commented on Classes/Processor.php:
```diff
@@ -312,4 +295,14 @@ public function getLocker(string $key): LockingStrategyInterface
+        $variants = explode(',', $accept);
+
+        return in_array($mimeType, $variants, true);
+    }
```
Resolved — determineBrowserSupport was removed entirely in ec0c1fb. The Processor now uses the main branch approach with query parameter-based variant selection.

CybotTM commented on Classes/Processor.php:
```diff
@@ -265,14 +250,12 @@ public function generateAndSend(): void
             $this->generateAvifVariant();
+        } elseif ($this->determineBrowserSupport('image/webp') && $this->isWebpImage() === false) {
+            $this->generateWebpVariant();
+        } else {
```
Resolved — the Processor was reverted to the main branch version which properly handles skipWebP/skipAvif query parameters. The ViewHelper and Processor are now in sync.

CybotTM commented on Classes/Command/OptimizeImagesCommand.php:
```diff
@@ -0,0 +1,269 @@
+            $progress->setMessage($this->shortenLabel($label, $messageMax));
+
+            $file    = $this->factory->getFileObject($record['uid']);
+            $ext     = $file->getExtension();
```
Fixed in 333532c — OptimizeImagesCommand now calls strtolower() on the extension and delegates to ImageOptimizer.resolveToolFor() which also normalizes.

CybotTM commented on AGENTS_QUICKSTART.md:
```diff
@@ -0,0 +1,179 @@
+├── Classes/AGENTS.md      ← PHP source code guidelines
+├── Tests/AGENTS.md        ← Unit testing guidelines
+├── Makefile               ← Developer commands (make help)
+├── .envrc                 ← Environment setup (direnv)
```
Resolved — both AGENTS_QUICKSTART.md and AGENTS_CHANGELOG.md have been removed from the branch.

CybotTM commented on Classes/Command/AbstractImageCommand.php:
```diff
@@ -0,0 +1,135 @@
+        }
+
+        return $qb->executeQuery()->fetchAllAssociative();
+    }
```
Fixed in 2a8235d — iterateViaIndex now uses a Generator with streaming iteration (fetchAssociative in a while loop) instead of fetchAllAssociative. A separate countImages() method provides the count for the progress bar.

CybotTM commented on Classes/Command/AnalyzeImagesCommand.php:
```diff
@@ -0,0 +1,107 @@
+        $total = [
+            'files'          => $count,
+            'improvable'     => 0,
+            'bytesPotential' => 0,
+            'noGain'         => 0,
+        ];
+
+        ['progress' => $progress, 'messageMax' => $messageMax] = $this->createProgress($io, $count);
+
+        foreach ($records as $record) {
+            $label = $this->buildLabel($record);
+            $progress->setMessage($this->shortenLabel($label, $messageMax));
+
+            $file = $this->factory->getFileObject($record['uid']);
+            $file->getStorage()->setEvaluatePermissions(false);
+
+            if (!$file->exists()) {
+                continue;
+            }
+
+            $result = $this->optimizer->analyzeHeuristic($file, $maxWidth, $maxHeight, $minSize);
+
+            if ($result['optimized']) {
+                ++$total['improvable'];
+                $total['bytesPotential'] += $result['savedBytes'];
+            } else {
+                ++$total['noGain'];
+            }
+
+            $progress->advance();
+        }
+
+        $progress->finish();
+        $io->newLine(2);
+
+        $io->success(sprintf('Analyse fertig. Dateien: %d, Verbesserbar: %d, Ohne Potenzial: %d, Potenzial: %d Bytes (%s)', $total['files'], $total['improvable'], $total['noGain'], $total['bytesPotential'], $this->formatMbGb($total['bytesPotential'])));
+
+        return self::SUCCESS;
+    }
+}
```
Acknowledged — unit tests for the new CLI commands and services will be added in a follow-up. The existing test infrastructure covers the Processor and ViewHelper.

CybotTM commented on Classes/Command/AnalyzeImagesCommand.php:
```diff
@@ -0,0 +1,107 @@
+
+#[AsCommand(
+    name: 'nr:image:analyze',
+    description: 'Analysiert Originale und ermittelt Optimierungspotenzial (optipng, gifsicle, jpegoptim).'
```
Fixed in f09f29e — the command description and all German UI strings in AnalyzeImagesCommand have been translated to English, and the description now accurately reflects that it uses heuristic analysis without external tools.

CybotTM committed:
style: align match arm whitespace in SourceSetViewHelper

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
feat(service): add ImageOptimizer for binary-driven image optimization

Shared service used by both the event listener and CLI commands. Wraps
optipng, gifsicle, and jpegoptim via Symfony Process, resolving binary
paths through \$PATH with env overrides (OPTIPNG_BIN, GIFSICLE_BIN,
JPEGOPTIM_BIN).

Provides:
- optimize()        — runs a tool in-place and replaces the original on shrink
- analyze()         — same pipeline but never writes back
- analyzeHeuristic()— fast estimation from size + resolution, no tool invocation
- resolveToolFor()  — exposes tool discovery for callers that want to pre-check
- hasAnyTool()      — quick check whether any optimizer is available

Extensions supported: png, gif, jpg, jpeg. Unsupported extensions return
early without invoking any binary.

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
feat(upload): auto-optimize images on file added/replaced events

Registers a PSR-14 listener against AfterFileAddedEvent and
AfterFileReplacedEvent. When a supported image (png/gif/jpg/jpeg) is
uploaded or replaced on an online storage, the listener asks
ImageOptimizer to run the appropriate tool in-place.

Safety:
- Re-entrancy guard prevents loops when the optimizer's own
  storage->replaceFile() call re-triggers AfterFileReplacedEvent.
- Storage permission checks are temporarily disabled (BE-independent
  context) and restored in a finally block so subsequent storage
  operations are unaffected.
- Unsupported extensions and offline storages short-circuit without work.

Also excludes Classes/Domain/Model/* from the catch-all service
registration to leave room for future value objects.

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
feat(cli): add AbstractImageCommand with shared FAL iteration helpers

Base class for the forthcoming image CLI commands. Provides:
- parseStorageUidsOption() — normalizes a repeatable --storages CLI arg
  (comma-separated or multi-occurrence) into list<int>
- getIntOption()           — numeric-safe option parsing with fallback
- countImages()            — COUNT(*) over indexed images, respecting
  storage filter and image/* mime types
- iterateViaIndex()        — Generator over sys_file rows so large
  installations don't load the full index into memory
- createProgress()         — SymfonyStyle progress bar + message width
- buildLabel()/shortenLabel()/extractUid() — UX helpers
- formatMbGb()             — human-friendly bytes->MB/GB rendering

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
feat(cli): add nr:image:optimize bulk optimization command

Walks the FAL index and optimizes png/gif/jpeg files across all TYPO3
storages, delegating to ImageOptimizer for the actual tool invocation.

Options:
- --dry-run        preview without writing
- --storages       restrict to specific storage UIDs (repeatable / CSV)
- --jpeg-quality   lossy JPEG quality (0-100); omit for lossless
- --strip-metadata remove EXIF/comments when the tool supports it

UX: SymfonyStyle progress bar with upfront record count, truncated
label messages that fit the terminal width, and MB/GB savings totals.
Per-file permission checks are disabled then restored in a try/finally
so long-running runs don't leak state across iterations.

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
feat(cli): add nr:image:analyze heuristic analysis command

Reports optimization potential without invoking any external tool or
modifying files. Uses a heuristic driven by file size and resolution
(configurable via --max-width, --max-height, --min-size), producing
an estimated savings total in bytes and MB/GB.

Fast even on large installations because it skips the process spawn
per file — useful for estimating the payoff before scheduling a full
nr:image:optimize run.

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
test: cover ImageOptimizer service

48 tests, 111 assertions. Covers:
- supportedExtensions()
- optimize()/analyze() early returns for unsupported extensions and
  missing tool binaries
- resolveToolFor() lowercase normalization and null for missing binaries
- hasAnyTool() with and without env-pinned binaries
- analyzeHeuristic() — all branches: unsupported ext, missing file,
  below-minSize, large-image scaling, getimagesize() fallback to
  file properties, missing dimensions, local-copy cleanup, per-type
  base gain
- buildJpegoptimArgs() via reflection — strip, quality clamping at
  both ends, zero, 100, negative

Tool binaries are pinned to /bin/true (no-savings path), /bin/false
(process failure), and a shell wrapper that truncates its argument
(shrink + Storage::replaceFile path) — no real optimizer is required.

clearstatcache() is called after padding temp files to avoid PHP's
stat cache returning the pre-pad size to production code.

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
test: cover OptimizeOnUploadListener

10 tests, 20 assertions. 100% line coverage. Covers:
- Offline-storage short-circuit
- Unsupported-extension short-circuit
- Uppercase extension normalization
- Re-entrancy guard (nested invocation during replaceFile)
- Permission disable/restore in finally — success and exception paths
- previousPermissions preservation (both true and false)
- Guard-cleared follow-up calls
- Different-identifier isolation

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
test: cover AbstractImageCommand helpers

55 tests, 55 assertions. 96% line coverage via a test fixture subclass
that exposes the protected helpers. Covers:
- parseStorageUidsOption: empty array, single UID, CSV, multiple values,
  non-numeric skipped, non-scalar skipped
- getIntOption: numeric string, int, null, non-numeric with default
- extractUid: missing key, numeric value, non-numeric, null
- buildLabel: with identifier, without, with non-scalar uid
- shortenLabel: short text unchanged, long text ellipsized, tiny maxLen edges
- formatMbGb: 0 bytes, 1 MB, 1 GB, large values
- createProgress: returns ProgressBar + messageMax int >= 10

countImages and iterateViaIndex depend on TYPO3 ConnectionPool and
are exercised via the command-level tests (which fake the connection).

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
test: cover OptimizeImagesCommand

10 tests. 96% line coverage. Covers execute() orchestration:
- hasAnyTool == false produces FAILURE exit
- Zero matching files produces SUCCESS with a warning
- Unsupported extensions are reported and counted as skipped
- Dry-run prints the plan without calling optimize()
- Shrinking optimization reports savings and invokes replaceFile
- Non-shrinking runs report "no savings"
- Process failures are reported per file
- --jpeg-quality option is parsed and forwarded
- Permissions are restored per iteration even when the tool fails

Uses GeneralUtility::addInstance() to fake ConnectionPool for the
final command class (which can't be subclassed). Real ImageOptimizer
with JPEGOPTIM_BIN env overrides drives each branch (/bin/true for
no-savings, /bin/false for failure, a truncating shell wrapper for
the shrink path).

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
test: cover AnalyzeImagesCommand

7 tests. 100% line coverage. Covers execute():
- Zero matching files produces SUCCESS with a warning
- Improvable files are counted and bytes aggregated
- Files where exists() is false still advance the progress bar
- Files below --min-size are classified as no-gain
- --max-width, --max-height, --min-size, --storages options are parsed
- Permissions are restored per iteration (finally block)

Uses the same GeneralUtility::addInstance() pattern as the optimize
command test to fake ConnectionPool without subclassing the final class.

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
chore(di): drop unused exclude for non-existent Classes/Domain/Model path

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
fix(upload): key re-entrancy guard by storage UID + identifier

FAL identifiers are only unique within a storage. Two different storages
can legitimately share an identifier like /image.jpg, so keying the guard
solely on the identifier would cause unrelated events to be skipped when
the same identifier appears in the same request against different
storages.

Composite key of storage UID + identifier prevents the cross-storage
collision while still catching the re-entrancy case that triggered this
guard (ImageOptimizer::optimize calls replaceFile, which re-fires
AfterFileReplacedEvent for the same storage+identifier).

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM committed:
fix(service): treat invalid *_BIN overrides as authoritative, verify executability

When OPTIPNG_BIN / GIFSICLE_BIN / JPEGOPTIM_BIN is set, it is now
authoritative: if the path doesn't exist or isn't executable, the
tool is reported unavailable. Previously the resolver fell back to
\`which\`, which silently masked misconfigured overrides and could run
a different binary than the operator expected.

Also verifies executability for the \`which\` result so a bare regular
file on PATH can't be invoked as a binary.

Tests add chmod 0755 on the fake binaries they write so the new
is_executable() check is satisfied.

Signed-off-by: Sebastian Mendel <info@sebastianmendel.de>

CybotTM commented on Configuration/Services.yaml:
```diff
@@ -6,6 +6,7 @@ services:
 
   Netresearch\NrImageOptimize\:
     resource: '../Classes/*'
+    exclude: '../Classes/Domain/Model/*'
```
Fixed in 6a0e18b — removed the unused exclude. The directory never existed and the line was just noise.

CybotTM commented on Classes/EventListener/OptimizeOnUploadListener.php:
```diff
@@ -0,0 +1,70 @@
+            return; // Re-entrancy guard (e.g. triggered by replaceFile)
+        }
+
+        $storage = $file->getStorage();
```
Fixed in e46053d — the guard key is now composed of storage UID + identifier, which prevents cross-storage collisions while still catching the replaceFile re-entrancy case that motivated the guard.

CybotTM commented on Classes/Service/ImageOptimizer.php:
```diff
@@ -0,0 +1,353 @@
+            return $path !== '' ? $path : null;
+        }
+
+        return null;
```
Fixed in 00c0dda — a set *_BIN env var is now authoritative: if the path is missing or not executable, the tool is reported unavailable (no silent PATH fallback). Also added is_executable() to the which result so a regular file on PATH can't be invoked as a binary.
